sds/src/scanner/test/metrics.rs

```diff
@@ -170,3 +170,123 @@ fn test_regex_match_and_included_keyword_same_index() {
+        .expect("cpu_duration metric not found");
+
+    // For non-async rules, CPU duration should be > 0
+    match &metric_value.2 {
+        DebugValue::Counter(value) => {
+            assert!(*value > 0, "CPU duration should be greater than 0");
+        }
+        _ => panic!("Expected Counter value"),
+    }
+}
+
+#[test]
+fn should_submit_cpu_duration_metric_with_async_rule() {
+    use crate::scanner::config::RuleConfig;
+    use crate::scanner::{CompiledRule, CreateScannerError, StringMatchesCtx};
+    use std::sync::Arc;
+
+    // Create a custom async rule that sleeps for 100ms
+    struct SleepyAsyncRuleConfig;
+
+    struct SleepyAsyncCompiledRule;
+
+    impl CompiledRule for SleepyAsyncCompiledRule {
+        fn get_string_matches(
+            &self,
+            _content: &str,
+            _path: &Path,
+            ctx: &mut StringMatchesCtx,
+        ) -> crate::scanner::RuleResult {
+            ctx.process_async(|_async_ctx| {
+                Box::pin(async move {
+                    // Sleep for 100ms to simulate I/O
+                    tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
+                    Ok(())
+                })
+            })
+        }
+    }
+
+    impl RuleConfig for SleepyAsyncRuleConfig {
+        fn convert_to_compiled_rule(
+            &self,
+            _rule_index: usize,
+            _labels: crate::Labels,
+        ) -> Result<Box<dyn CompiledRule>, CreateScannerError> {
+            Ok(Box::new(SleepyAsyncCompiledRule))
+        }
+    }
+
+    let recorder = DebuggingRecorder::new();
+    let snapshotter = recorder.snapshotter();
+
+    metrics::with_local_recorder(&recorder, || {
+        let rule_0 = RootRuleConfig::new(Arc::new(SleepyAsyncRuleConfig) as Arc<dyn RuleConfig>)
+            .match_action(MatchAction::None);
+
+        let scanner = ScannerBuilder::new(&[rule_0]).build().unwrap();
+        let mut content = SimpleEvent::Map(BTreeMap::from([(
+            "key1".to_string(),
+            SimpleEvent::String("test content".to_string()),
+        )]));
+
+        // Use scan (which blocks on async internally)
+        scanner.scan(&mut content).unwrap();
+    });
+
+    let snapshot = snapshotter.snapshot().into_hashmap();
+
+    let metric_name = "scanning.cpu_duration";
+    let metric_value = snapshot
+        .get(&CompositeKey::new(Counter, Key::from_name(metric_name)))
+        .expect("cpu_duration metric not found");
+
+    // CPU duration should be much less than 100ms since we slept during I/O
+    match &metric_value.2 {
+        DebugValue::Counter(value) => {
+            // CPU duration should be < 10ms (10_000_000 nanoseconds)
+            // Since we slept for 100ms, the actual CPU time should be minimal
+            assert!(
+                *value < 10_000_000,
+                "CPU duration should be less than 10ms, got {} ns",
+                value
+            );
+        }
```

Comment:
I think it could be a bit flaky when the CI is slow. An other option would be to measure the duration before and after  ` scanner.scan(&mut content).unwrap();` and validate that this duration is at least 100ms bigger than the metric value